Update CLI

mustafasoylu opened the pull request:
- Updated CLI options
- Added new tests to increase coverage to above 95%
- Fixed some minor bugs thanks to additional tests
- Added .somesy.toml input support
- Using logging to support different levels of information output
- pre-commit settings
- updated documentation

mustafasoylu commented:
@apirogov, the PR is ready for review with this commit.

apirogov commented on src/somesy/main.py:
```diff
@@ -1,4 +1,138 @@
+            f"CLI arguments:\n{input_file=}, {no_sync_cff=}, {cff_file=}, {no_sync_pyproject=}, {pyproject_file=}, {verbose=}, {debug=}"
+        )
+
+        # at least one of the sync options must be False
```
you mean true? I get it ... the flag is a negation.... maybe write `must be enabled` :D

mustafasoylu replied:
I changed this part a lot; deciding how to have it in CLI options was hard. So, I don't even know what I meant at the first time when I wrote this comment. I will update it accordingly :)

apirogov commented on src/somesy/main.py:
```diff
@@ -1,4 +1,138 @@
+        raise typer.Exit(code=1)
+
+
+if __name__ == "__main__":
+    app()
```
I don't think you actually need this. You can't run a project without installing it that uses `src` layout anyway ;)

mustafasoylu replied:
I don't know why, but it created a problem in testing the typer CLI commands. This is why I put it this way.

apirogov commented on src/somesy/core/utils.py:
```diff
@@ -0,0 +1,48 @@
+        if self.isEnabledFor(VERBOSE):
+            self._log(VERBOSE, message.format(args), (), **kwargs)
+
+    setattr(logging.Logger, "verbose", verbose_print)  # noqa: B010
```
why not use a `wrapt` objectproxy and add the new method, passing everything else through?

then you won't need the type ignores at call sites and its clean.

mustafasoylu replied:
@apirogov created [this issue](https://github.com/Materials-Data-Science-and-Informatics/somesy/issues/4 ) to refactor this function later.

apirogov commented on src/somesy/main.py:
```diff
@@ -1,4 +1,138 @@
+        )
+
+        # at least one of the sync options must be False
+        if no_sync_cff and no_sync_pyproject:
+            raise ValueError("There should be at least one file to sync.")
```
Is that actually a problem? I'd rather log a warning. Furthermore, this check duplicates the check in the actual sync method. I'd use the typer-decorated method to kind of pre-parse the arguments, but push the actual validation logic like that down into the "library function" and not duplicate it.

mustafasoylu replied:
Removed check in the command. Now it checks these two flags first and gives a warning and exits

apirogov commented on src/somesy/main.py:
```diff
@@ -1,4 +1,138 @@
+        # sync files
+        sync_command(
+            input_file=input_file,
+            pyproject_file=Path(pyproject_file) if not no_sync_pyproject else None,
+            cff_file=Path(cff_file) if not no_sync_cff else None,
```
aren't the arguments already `Path` objects? the type hint is `Path` so `typer` should give you a path? then you can also get rid of the `if`s

mustafasoylu replied:
I remember it was passed as a string. This is why is do dynamic recasting with Path

apirogov commented on src/somesy/main.py:
```diff
@@ -1,4 +1,138 @@
+            raise ValueError("There should be at least one file to sync.")
+
+        # check if input file exists, if not, try to find it from default list
+        input_file = discover_input(str(input_file) if input_file else None)
```
Why do you convert this to a string, just to convert is back to a `Path` in the method?

If there's no reason maybe change `discover_input` to take a `Path` and not a `str`, and just pass it?

then you also don't need the inline `if` for the `None` case (its optional anyway)

apirogov commented on src/somesy/core/core.py:
```diff
@@ -10,32 +10,38 @@ def get_project_metadata(path: Path) -> ProjectMetadata:
-        try:
-            with open(path, "r") as f:
-                config = load(f)
-        except FileNotFoundError:
-            raise FileNotFoundError("pyproject.toml not found")
+        with open(path, "r") as f:
+            input = load(f)
 
         # load project metadata
-        if (
-            "tool" in config
-            and "somesy" in config["tool"]
-            and "project" in config["tool"]["somesy"]
+        if "somesy" in path.name and "project" in input:
+            metadata = input["project"]
+            try:
+                metadata = metadata.unwrap()
+                return ProjectMetadata(**metadata)
+            except Exception as e:
+                raise ValueError(f"Somesy input validation failed: {e}")
+
+        elif (
+            "pyproject" in path.name
+            and "tool" in input
+            and "somesy" in input["tool"]
+            and "project" in input["tool"]["somesy"]
         ):
-            metadata = config["tool"]["somesy"]["project"]
+            metadata = input["tool"]["somesy"]["project"]
             try:
                 metadata = metadata.unwrap()
                 return ProjectMetadata(**metadata)
             except Exception as e:
                 raise ValueError(f"Somesy input validation failed: {e}")
         else:
-            raise ValueError("Somesy config not found in pyproject.toml file.")
+            raise ValueError("Somesy input not found.")
     else:
-        raise ValueError("Only toml files are supported for reading project metadata.")
+        raise ValueError("Only toml files are supported for reading somesy input.")
```
Phew, this looks very convoluted, and like something you'd need to touch whenever you add a new input source. And its not easy to understand, because all the input types are entangled. I get what you're doing, but I'm not really happy about this. Well, maybe something to keep in mind and refactor in a subsequent iteration.

mustafasoylu replied:
@apirogov what is your recommendation?

mustafasoylu replied:
@apirogov created [this issue](https://github.com/Materials-Data-Science-and-Informatics/somesy/issues/5). I will do it at a later iteration.

apirogov reviewed (commented):
I added some comments :)

mustafasoylu commented on .gitignore:
```diff
@@ -1,5 +1,6 @@
 # don't add custom python shell scratchpad file
 scratch*.py
+scratch
```
I have a whole folder with that name :)

mustafasoylu commented on .pre-commit-hooks.yaml:
```diff
@@ -0,0 +1,7 @@
+  description: Sync your metadata files with somesy
+  entry: somesy sync
+  language: python
+  always-run: true
```
makes sense, updating it

mustafasoylu commented on src/somesy/commands/sync.py:
```diff
@@ -0,0 +1,79 @@
+    if cff_file is not None:
+        _sync_cff(metadata, cff_file, create_cff)
+
+    if pyproject_file is None and cff_file is None:
+        raise ValueError("At least one of the output files must be specified.")
```
I will move it to the beginning. Since it is an error, it will be logged as error in the main typer command.

mustafasoylu commented on src/somesy/core/discover.py:
```diff
@@ -0,0 +1,39 @@
+    """
+    if input_file:
+        p = Path(input_file)
+        if p.exists():
```
makes sense!

mustafasoylu commented on src/somesy/pyproject/core.py:
```diff
@@ -29,26 +32,32 @@ def __init__(self, path: Path, create_if_not_exists: bool = False):
                 path.touch()
                 # if empty file is created, add tool.poetry object
-                dump({"tool": {"poetry": {}}}, path)
+                with open(path, "w") as f:
+                    dump({"tool": {"poetry": {}}}, f)
```
I was unsure if we should have this as a feature. Currently, CLI does not have the option for this; I will explain in the documentation that you have to have a valid pyproject.toml file if you want to sync it. Thus, this part of the code is redundant. I may remove these parts for the pyproject class.

mustafasoylu replied:
Removed creating a new poetry or setuptools file if it does not exist. It would cause problems in the long run.

mustafasoylu commented on src/somesy/cff/core.py:
```diff
@@ -55,7 +55,7 @@ def _load(self):
         """Validate the CFF file."""
         if not validate_citation(self.path):
-            raise Exception("CFF file is invalid")
+            raise ValueError("CFF file is invalid")
```
Updated the cff-convert error message with debug option

mustafasoylu commented:
@apirogov I took all of your comments into consideration and made the necessary updates. I created two issues for refactoring two parts, which I will get into later. I also updated AUTHORS.md, .somesy.toml, and README.md. I added CLI options explanation and examples of somesy inputs in README.md. Looking forward to your final review.

apirogov reviewed (commented):
Nice, looks good. I'll give it a test run later (I'll try adding it to dirschema) and report back, if it works you can then go ahead ;)

mustafasoylu commented on pyproject.toml:
```diff
@@ -25,10 +13,11 @@ documentation = "https://materials-data-science-and-informatics.github.io/somesy
 classifiers = [
     "Operating System :: POSIX :: Linux",
-    "Development Status :: 2 - Pre-Alpha",
+    "Development Status :: 5 - Production/Stable",
```
I was not sure. Actually, I will just delete it for now.

apirogov commented on .pre-commit-hooks.yaml:
```diff
@@ -0,0 +1,8 @@
+  description: Sync your metadata files with somesy
+  entry: somesy sync
+  language: python
+  always-run: true
```
I get 

```
[WARNING] Unexpected key(s) present on https://github.com/Materials-Data-Science-and-Informatics/somesy => somesy: always-run
```

And also, shouldn't this be removed? (that was the reason for adding the `files` key)

apirogov commented:
I get

```

Error: 1 validation error for PoetryConfig                                      
plugins -> dirschema_validator                                                  
  str type expected (type=type_error.str)
```

error when I try to use it :(

mustafasoylu commented:
> I get
> 
> ```
> 
> Error: 1 validation error for PoetryConfig                                      
> plugins -> dirschema_validator                                                  
>   str type expected (type=type_error.str)
> ```
> 
> error when I try to use it :(

@apirogov

I made the necessary update with [the last commit](https://github.com/Materials-Data-Science-and-Informatics/somesy/pull/2/commits/de8ef7298427692051adcd75f6cda93bf9c2b9cc). Could you try again?

apirogov reviewed (approved):
Apart from the issues with nested metadata (inside a person subobject) I guess it's fine :+1:

the issues were
* not preserving of nested information e.g. affiliation
* not preserving formatting of such fields
